### ISSUE TITLE
Filter by Array of Primitives

### ISSUE THREAD

**Issue by tomitrescak** [prisma/photonjs#341]
Hi, in prisma/prisma2#169 you have implemented the support for Array of primitives.
Do you have a plan for supporting filtering by such array?

For example:

```
model Person {
   titles: String[]
}

person.findMany({ where: { titles: { contains: 'Dr' } } })
person.findMany({ where: { titles: { in: ['Dr', 'Ing'] } } })
```

**Comment by tre-dev** [comment]
Hm, this is quite an important feature for me. Are you able to ship this with Q1 Production release? Otherwise, I'll probably need to rethink how I deal with this problem, as it's pretty much a requirement that I'm able to query/filter based on arrays. Is there an alternative/workaround?

**Comment by BjoernRave** [comment]
Not sure if this also includes being able to filter for multiple enums, but that's something I would need

**Comment by MJGTwo** [comment]
This is an important feature for me, especially since we had something similar in Prisma 1. If I have the datamodel: 
```
model User {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  email     String   @unique
  name      String?
  role      Role     @default(USER)
  posts     Post[]
}

model Post {
  id         Int        @id @default(autoincrement())
  createdAt  DateTime   @default(now())
  updatedAt  DateTime   @updatedAt
  title      String
  published  Boolean    @default(false)
  author     User
}

enum Role {
  USER
  ADMIN
}
```

I want to be able to do a `findMany` on unique and non-unique fields with an array of values in order to get an array of results. Such as:
```
const emails = [alice@example.com, bob@example.com];
user.findMany({ where: { emails: { in: emails } } });
//
const titles = ['dr', 'rev', 'capt'];
user.findMany({ where: { titles: { not: titles } } });

```

Currently it looks like I have to daisy-chain with `OR`,  and `NOT` if there's an exact number of criteria (poor coding) or do `await` and `async` with a `.map` which leads to decreased performance from multiple queries happening against each other.

**Comment by PetrVasilev** [comment]
For me is very important feauture!!!

**Comment by esistgut** [comment]
Our prisma based ecommerce product currently has a couple of features blocked by this issue and we would really like to avoid reverting to the `queryRaw` route. Is there any ETA?
(This is a "+1 comment", I don't usually do this but I've been instructed on Slack to do so).

**Comment by beeman** [comment]
Also leaving a comment here to notify I'm currently blocked by the lack of having this feature. Like @esistgut I'd love to not fall back to using `queryRaw`. 

It would be lovely if the Prisma product team could prioritize and give us an ETA so we set expectations for our team and clients.

**Comment by mavilein** [comment]
During the last planning we were unsure whether this is already partially implemented or not. I found some corresponding data structures in our internal filter API. But those are not implemented in any way. 
So the following is still needed:
1. Define what the API should look like in the client.
2. Define an implementation for each operation in the API. As only Postgres supports scalar lists right now here is the link to the [docs](https://www.postgresql.org/docs/12/functions-array.html) for their array functions.

Internally we have the following filters defined for scalar lists right now. Are these already enough? Do we want more?
```
pub enum ScalarListCondition {
    /// List contains the given value
    Contains(PrismaValue),

    /// List contains all the given values
    ContainsEvery(Vec<PrismaValue>),

    /// List contains some of the given values
    ContainsSome(Vec<PrismaValue>),

    /// List is empty
    ContainsNone,
}
```

**Comment by thebiglabasky** [comment]
Thanks @mavilein! These looks like definitely enough for now.
I would let @timsuchanek chime in for the client API piece

**Comment by mavilein** [comment]
@do4gr mentioned that we might want an equality condition additionally.

**Comment by Sytten** [comment]
@thebiglabasky This is still tagged 2.2.0 but is not implemented as far as I am aware, it is actually missing. Is it possible to put it back in the roadmap somewhere? Thanks

**Comment by thebiglabasky** [comment]
Hi,
This is still in the "Planned" state right now so coming in the next months. We're busy going after native types and atomic operations first.

**Comment by Sytten** [comment]
@mavilein we also need the inverse operation too: `ContainedIn: Vec<PrismaValue>` where the input `vec` contains the stored `vec`.

**Comment by thebiglabasky** [comment]
@MJGTwo what you describe is currently supported.
`String` fields are capable of taking `in`, `notIn` etc.
So in your example
```
const emails = [`alice@example.com`, `bob@example.com`];
user.findMany({ where: { emails: { in: emails } } });
//
const titles = ['dr', 'rev', 'capt'];
user.findMany({ where: { titles: { not: titles } } });
```
Will work as expected.

The current limitation applies to `String[]` or any other field for which the exact database type is an array of scalar. The latter is only supported by Postgres though.

I created an issue in the prisma/prisma repo to summarize the problem and suggested solutions: https://github.com/prisma/prisma/issues/3475
I would suggest people to track this one instead

**Comment by albertoperdomo** [comment]
Additional context/request in https://github.com/prisma/prisma-client-js/issues/334.

**Comment by narciero** [comment]
any update on when we can expect this feature? thanks!

**Comment by thebiglabasky** [comment]
We're wrapping up group by this sprint, so this is something that we'll have to look at after the holidays, early next year.

**Comment by matthewmueller** [comment]
Hey folks, here's the current plan for supporting filtering scalar list fields. 

Please let me know if you have any thoughts, concerns or suggestions!

You can also schedule time with me to discuss privately: https://calendly.com/matthewmueller/prisma

---

## Functionality

### Databases

- Postgres

### Supported Operations

- findFirst
- findMany
- groupBy
- aggregate

### Capabilities

We add the following has, hasEvery, hasSome and isEmpty as filters for scalar lists. These capabilities can be combined with our NOT, OR, and AND operations.

- **has:** a value is contained within a list
    - [1, 2, 3] has 1 is **true**
    - [1, 2, 3] has 4 is **false**
- **hasEvery:** all values are contained within the list
    - [1, 2, 3] hasEvery [1, 2] is **true**
    - [1, 2, 3] hasEvery [1, 2, 4] is **false**
- **hasSome:** at least one values is contained in the list
    - [1, 2, 3] hasSome [1, 2] is **true**
    - [1, 2, 3] hasSome [1, 2, 4] is **true**
    - [1, 2, 3] hasSome [4, 5] is **false**
- **isEmpty:** the list is empty
    - [] is empty is **true**
    - [1] is empty is **false**

### Supported Types

We'll start by trying to add all the types, see how the tests react and adjust our expectations accordingly.

**Must Haves**

- String[]
- Enum[]

**Nice to Haves**

- Int[]
- Float[]
- Boolean[]
- DateTime[]

**Icing on the Cake**

- BigInt[]
- Decimal[]

### Out of Scope

- Json[]
- Bytes[]
- Postgres native types and extensions (e.g. uuid, citext, postgis)

## Example

Given the following Schema:

```groovy
model User {
  id     Int      @id
  active Boolean
  roles  String[]
  titles Title[]
  level  Int[]
}

enum Title {
  Dr
  Ing
  Esq
}
```

You should be able to query:

```jsx
*// Read: Find first user who has a "Dr" title*
await prisma.user.findFirst({
  where: {
    titles: {
      // "title" contains the given value
      has: "Dr"
    }
  }
})

// *Read: Find first user who has both a "Dr" and "Ing" Title*
await prisma.user.findFirst({
  where: {
    titles: {
      // "title" contains all the given values.
      // In this case both "Dr" and "Ing"
      hasEvery: ["Dr", "Ing"]
    }
  }
})

// *Read: Find first user who either a "Dr" or a "Ing" Title*
await prisma.user.findFirst({
  where: {
    titles: {
      // "title" contains at least one of the given values.
      // In this case "Dr" or "Ing"
      hasSome: ["Dr", "Ing"]
    }
  }
})

// *Read: Find first user who has no titles.*
await prisma.user.findFirst({
  where: {
    titles: {
      // "title" is empty
      isEmpty: true // or hasNone
    }
  }
})

// These filter clauses (contains, some, every, none) should work for 
// findFirst, findMany & findUnique. (TODO: aggregate, groupBy too?)
// 
// They should also work with other where clauses. The operations are 
// ANDed.
//
// Read: find all active users that have either a Dr or Ing title.
await prisma.user.findMany({
  where: {
    active: true
    titles: {
      hasSome: ["Dr", "Ing"]
    }
  }
})

// Works with explicit OR clauses too.
//
// Read: find all active users that have either a Dr. or Ing. title 
// OR don't have any titles.
await prisma.user.findMany({
  where: {
    active: true,
    OR: [
      titles: {
        hasSome: ["Dr", "Ing"]
      },
      titles: {
        isEmpty: true
      },  
    ]
  }
})

// Lastly, everything above should also work for Enum and Int types. 
await prisma.user.findMany({
  where: {
    active: true,
    OR: [
      roles: {
        hasSome: ["email:read", "user:write"]
      },
      roles: {
        hasNone: true
      },  
    ]
  }
})
```

**Comment by Sytten** [comment]
Couple of observations:
- `Json[]` might be out of scope, but we still have to think how we will be able to compare elements inside the array when its going to be objects so it won't be necessary to do a breaking change later on.
- Will it work if we have `ids String[]` where the underlining storage type is UUID[]?
- One nice to have would be `hasNone` which is basically a `NOT: { field: hasSome(...) }`, but easier to read.

**Comment by nrxus** [comment]
With the proposal above, how would one check if two arrays are equal? 

`hasEvery` allows for the array in the DB to have more values than the array in the filter. In my use case equality check of two arrays is what I need.

**Comment by Sytten** [comment]
I think the default behaviour that already exists is strict equality @nrxus

**Comment by nrxus** [comment]
Oh I could have swore that wasn't there a few versions ago. Thanks!
